### ISSUE TITLE
Should it be a mean or a sum in the loss function of the variational autoencoder example?

### ISSUE THREAD

**Issue by wagglefoot** [fchollet/keras#6015]
In the `variational_autoencoder.py` example, on line 46 you use:

```
kl_loss = - 0.5 * K.sum(1 + z_log_var - K.square(z_mean) - K.exp(z_log_var), axis=-1)
```
 [but in your blog](https://blog.keras.io/building-autoencoders-in-keras.html), you use 
```
kl_loss = - 0.5 * K.mean(1 + z_log_sigma - K.square(z_mean) - K.exp(z_log_sigma), axis=-1).
```

Hence, should there be a `mean` or a `sum` in the loss function?

**Comment by LexSong** [comment]
They are not very different.

In my experiments, I try to add different weights on `x_loss` and `kl_loss`, still able to optimize.
In your example, `sum` and `mean` are just give different weights to `kl_loss`.

**Comment by Ultramann** [comment]
I think that both will work, but the version in `variational_autoencoder.py` appears to be more in line with the [paper](https://arxiv.org/pdf/1312.6114.pdf) cited at the top of the script. How, you ask, can both work? Notice the difference in the definition of `xent_loss`. In the blog `binary_crossentropy` is used in an unscaled form, and `mean` is used for `kl_loss`. In the script `binary_crossentropy` is scaled by a factor of `original_dim`, effectively undoing the averaging across all the pixels that `binary_crossentropy` is doing under the hood. This matches the use of `sum` in the script. Thus, the magnitude of `xent_loss` is scaled by the same factor as `kl_loss`. Since this is going to be a constant factor it won't affect the gradient.

**Comment by greatblueheron** [comment]
Ultramann: I don't think that is correct. The sum in the KL term is summing over the latent dimension, which is not the same as `original_dim`. I believe the correct form of the VAE loss has the first term scaled by `original_dim`, with the KL term being the sum.
